apis/r/R/SOMASparseNDArray.R

```diff
@@ -181,15 +189,112 @@ SOMASparseNDArray <- R6::R6Class(
         self$tiledb_timestamp %||% "now"
       )
 
-      private$.write_coo_dataframe(coo)
+      self$.write_coordinates(coo)
+      # private$.write_coo_dataframe(coo)
 
-      invisible(self)
+      return(invisible(self))
     },
 
     #' @description Retrieve number of non-zero elements (lifecycle: maturing)
     #' @return A scalar with the number of non-zero elements
     nnz = function() {
       nnz(self$uri, private$.soma_context)
+    },
+
+    #' @description Write a COO table to the array
+    #'
+    #' @param values A \code{data.frame} or \code{\link[arrow:Table]{Arrow::Table}}
+    #' with data in COO format; must be named with the dimension and attribute
+    #' labels of the array
+    #'
+    #' @return Invisibly returns \code{self}
+    #'
+    .write_coordinates = function(values) {
+      private$check_open_for_write()
+      dnames <- self$dimnames()
+      attrn <- self$attrnames()
+
+      stopifnot(
+        "'values' must be a data frame or Arrow Table" = is.data.frame(values) ||
+          inherits(values, what = 'Table'),
+        "'values' must have one column for each dimension and the data" = ncol(values) == length(dnames) + 1L,
+        "'values' must be named with the dimension and attribute labels" = is.null(names(values)) ||
+          identical(names(values), c(dnames, attrn))
+      )
+
+      # Arrow Tables cannot have NULL names, so this only applies to data frames
+      if (is.null(names(values))) {
+        spdl::warn("[SOMASparseNDArray$.write_coordinates] no names on input data frame, assuming <dimensions[...], data> order")
+        names(values) <- c(dnames, attrn)
+      }
+
+      # Check dimensions
+      spdl::debug("[SOMASparseNDArray$.write_coordinates] checking dimension values")
+      for (i in seq_along(dnames)) {
+        dn <- dnames[i]
+        integerish <- if (is.data.frame(values)) {
+          rlang::is_integerish(values[[dn]], finite = TRUE) ||
+            (bit64::is.integer64(values[[dn]]) && !any(is.na(values[[dn]])))
+        } else {
+          grepl('^int[[:digit:]]{1,2}$', x = values[[dn]]$type$name)
+        }
+        offending <- sprintf("(offending column: '%s')", dn)
+        if (!integerish) {
+          stop("All dimension columns must be integerish ", offending)
+        }
+        if (as.logical(min(values[[dn]]) < 0L)) {
+          stop("Dimension columns cannot contain negative values ", offending)
```

Comment:
perf implications of doing these full data scans, tradeoff against error-checking?

Reply:
My thought process is
 - only checking one value (`min()`/`max()`), pulling this value is fast and a single check is fast
 - most users will go through `$write()`, which is limited to 2^32 - 1 non-zero values due to limitations of captial-m-Matrix
 - `$.write_coordinates()` could support more, but with the introduction of my DelayedArray front-end, most usage of this from outside of tiledbsoma-r will be through chunks from DelayedArray (which must fit in memory)